### PR TITLE
add dpo length normalization

### PR DESCRIPTION
**Pull request by jacklanchantin**
**What does this PR do? Please describe:**
Adds an argument and functionality to use a length-normalized version of [DPO](https://arxiv.org/abs/2305.18290), by using the average log probability of a sequence as the implicit reward, as is done in [TÜLU 3](https://arxiv.org/abs/2411.15124).

Tested using this in my config:

```
criterion: dpo
_add_:
  criterion_config:
    beta: 0.1
    nll_scale: 0.0
    length_normalization: true
```

image

**Does your PR introduce any breaking changes? If yes, please list them:**
No

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
       - there were some failed tests in tests/unit/data/audio/test_waveform_to_fbank_converter.py
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/recipes/lm/preference_finetune/dpo.py [hunk 1]
```diff
@@ -116,6 +136,19 @@ def __call__(self, batch: PreferenceOptimizationBatch) -> tuple[Tensor, int]:
...
+        logprobs = torch.log_softmax(output.logits, dim=-1)
+        per_token_logps = torch.gather(logprobs, -1, target.seqs.unsqueeze(-1)).squeeze(
+            -1
+        )
+        total_logps = (per_token_logps * target.target_mask).sum(dim=-1)  # [Batch, 1]
+        assert target.target_mask is not None
+        average_logps = total_logps / target.target_mask.sum(-1)
+
+        return total_logps, average_logps
+
```
[the remark this answers is not in the document]
Comment (jacklanchantin) [reply src/fairseq2/recipes/lm/preference_finetune/dpo.py:152]: https://github.com/facebookresearch/fairseq2/pull/862/commits/3e07a7dd4c929f9b01d5b6c9c31065bf13dbaedb

**Review by uralik — APPROVED** [review]
thank you!
